src/backend/distributed/executor/local_executor.c

```diff
@@ -560,14 +561,26 @@ ExecuteLocalTaskPlan(PlannedStmt *taskPlan, char *queryString,
 	QueryEnvironment *queryEnv = create_queryEnv();
 	int eflags = 0;
 	uint64 totalRowsProcessed = 0;
+	bool storeTuples = true;
 
 	RecordNonDistTableAccessesForTask(task);
 
+	/*
+	 * For modifications to reference tables (or any replicated table in
+	 * general), remote execution have already stored the tuples and
+	 * incremented es_processed, so skip now.
+	 */
+	if (task->partiallyLocalOrRemote && !ReadOnlyTask(task->taskType) &&
```

Comment:
we already have tests with reference tables that triggers  this code=path. It'd be nice to have similar tests with replication > 1, one placement on  the coordinator the other is on the worker.

Reply:
Why do we have the !ReadOnlyTask distinction?

Reply:
Maybe results should always come from local execution for modifications as well

Reply:
@marcocitus as we talked, now whenever the task is `partiallyLocalOrRemote`, we store the tuples from the local execution.

Well, it turns out it has always been the case. But, due to the bug in `ReceiveResults ()`, we ended up with weird hacks. Please consider the change in `ReceiveResults ()` as well.

Overall, the code seems much cleaner to me now. Let's see what you think.